File: test/conformance/dart/rows/W8.dart
// W8 — bytes(N) takes the array row (the dart leg, fixed form, form byte 3).
//
// THE LAW (docs/FIXED-FORM-ALGORITHM.md:262, §4.1, fix 10): "A `bytes(N)` takes
// the ARRAY's row, not a text field's — the count to `aux` and the elements to
// `dst`", where a text field's row is the other way round — dst the length,
// aux the buffer. `bytes(N)` rides as an array of u8 on this wire (§3.4, kind
// 14), so under the text convention a plan compiled from a bytes(N)'s row
// hands the compiler a count destination that is the BUFFER'S FIRST FOUR BYTES
// and an element destination that is the LENGTH FIELD. The write side rides
// with it (§3.1, docs/FIXED-FORM-ALGORITHM.md:202): text and bytes write
// `length` units, never all Max.
//
// THE PRODUCTION PATH this test drives, end to end, every stage generated:
//
//   profileConfigFixedSave      the write: the template, then the stores
//                               (internal/codegen/darttable/fixeddart.go,
//                               emitWriteBody/emitWritePayload)
//   TableFixedCompiler.compile  the plan compiler reading MY dst rows
//                               (internal/codegen/darttable/fixedruntime.go,
//                               compileEntry, case 14 — the row this law is)
//   tableFixedRun               the one read loop over the compiled plan
//   profileConfigFixedDecode    the decode every path shares
//
// THE VECTOR is built here, from the law, because testdata/conformance/tables
// carries no fixed-form vector for this cell (its fixtures are the block, cook
// and json surfaces): one record whose bytes(16) field `icon` carries FOUR
// live bytes — the writer writes `length` units, never all Max, so the field's
// slack past the live byte stays the template's zeros. The offsets come from
// the declaration (tables/examples/Tables.schema, ProfileConfig): `name
// string(32)` takes 4 + 32 = 36 bytes, so icon's LENGTH is at body offset 36
// and its BUFFER at 40 — the same offsets profileConfigFixedDecode reads (at +
// 36, at + 40) and the identity plan's text entry carries (dst 36, aux 40).
//
// THE CONTROL is the old row, built by swapping the two columns back on a
// copy of this build's own rows and compiling the same plan from the same
// layout — the twin of test/tables/fixedform_main.cpp bytes_row_case, which
// is the C++ reference's assertion of this cell. The row is found by its shape
// and not by its index, so the control does not quietly stop pointing at it
// the day a field moves.
//
//   dart run test/conformance/dart/rows/W8.dart      (from the repo root)
//
// Exit 0 green, 1 red; one line per assertion.

import 'dart:io';
import 'dart:typed_data';

import '../../../../build/tables-generated-dart/examples/TablesFixed.dart'
    as tables;
import '../../../../build/tables-generated-dart/examples/TabledemoFixed.dart'
    as home;

int failures = 0;

void check(bool ok, String what) {
  stdout.writeln('${ok ? "ok" : "FAIL"} - $what');
  if (!ok) {
    failures++;
  }
}

// THE VECTOR, from the law: four live bytes, length 4, never all Max.
final List<int> live = <int>[0xDE, 0xAD, 0xBE, 0xEF];
const int iconLengthAt = 36; // 4 + 32: name string(32) before icon
const int iconBufferAt = 40;

void main() {
  // ---- THE ROW ------------------------------------------------------------
  //
  // The five lanes of a dst row: dst, stride, aux, counted, arg. The bytes row
  // is found by its SHAPE — stride one and a live count is a bytes(N) and
  // nothing else that also carries the bytes flavour — and not by its index.
  final rows =
      tables.profileConfigFixedDst.length ~/ home.TableFixedLane.dstLanes;
  var bytesRow = -1;
  var bytesRows = 0;
  var textRow = -1;
  var textRows = 0;
  for (var r = 0; r < rows; r++) {
    final b = r * home.TableFixedLane.dstLanes;
    final counted =
        tables.profileConfigFixedDst[b + home.TableFixedLane.dstCounted];
    final stride =
        tables.profileConfigFixedDst[b + home.TableFixedLane.dstStride];
    final arg = tables.profileConfigFixedDst[b + home.TableFixedLane.dstArg];
    if (stride == 1 && counted != 0 && arg == home.TableFixedOp.textBytes) {
      bytesRow = r;
      bytesRows++;
    }
    // a utf8 TEXT row: the flavour 1, no live count, no stride
    if (stride == 0 && counted == 0 && arg == 1) {
      textRow = r;
      textRows++;
    }
  }
  check(
    bytesRows == 1,
    'W8: the bytes(N) row is the one row with stride one, a live count and the '
    'bytes flavour (found $bytesRows)',
  );
  check(
    textRows == 1,
    'W8: the utf8 text row is the one row with no count, no stride and the '
    'utf8 flavour (found $textRows)',
  );

  // THE ROW ITSELF: the count to `aux` (the length field at 36) and the
  // elements to `dst` (the buffer at 40) — the ARRAY's row. The identity plan
  // lands this field with the TEXT op (length 36, units 40) and reads neither
  // column; the compile row is what a stranger's plan is walked through.
  if (bytesRow >= 0) {
    final b = bytesRow * home.TableFixedLane.dstLanes;
    check(
      tables.profileConfigFixedDst[b + home.TableFixedLane.dstAux] ==
          iconLengthAt,
      'W8: the bytes(N) row sends the count to aux — the length field at '
      '$iconLengthAt (got ${tables.profileConfigFixedDst[b + home.TableFixedLane.dstAux]})',
    );
    check(
      tables.profileConfigFixedDst[b + home.TableFixedLane.dstOffset] ==
          iconBufferAt,
      'W8: the bytes(N) row sends the elements to dst — the buffer at '
      '$iconBufferAt (got ${tables.profileConfigFixedDst[b + home.TableFixedLane.dstOffset]})',
    );
  }
  // ...AND A TEXT FIELD'S ROW IS THE OTHER WAY ROUND: dst the length (0),
  // aux the buffer (4) — name string(32) is the field before icon.
  if (textRow >= 0) {
    final b = textRow * home.TableFixedLane.dstLanes;
    check(
      tables.profileConfigFixedDst[b + home.TableFixedLane.dstOffset] == 0 &&
          tables.profileConfigFixedDst[b + home.TableFixedLane.dstAux] == 4,
      'W8: a text field\'s row is the other way round — dst the length (0), '
      'aux the buffer (4)',
    );
  }

  // ---- THE WRITE, THEN THE READ THROUGH THE COMPILED PLAN -----------------
  //
  // The record: icon carries four live bytes; the writer puts the length word
  // at 36 and those four bytes at 40, and nothing past them.
  final one = home.ProfileConfig();
  one.icon.setRange(0, live.length, live);
  one.iconLength = live.length;
  final file = Uint8List(tables.profileConfigFixedMeasure(1));
  check(
    tables.profileConfigFixedSave(<home.ProfileConfig>[one], 1, file) ==
        file.length,
    'W8: the production writer saves the record',
  );
  final bodyAt = tables.profileConfigFixedHeaderBytes + 8;
  final fileView = ByteData.sublistView(file);
  check(
    fileView.getInt32(bodyAt + iconLengthAt, Endian.little) == live.length,
    'W8: the length word rides in front, at $iconLengthAt',
  );
  var liveOk = true;
  for (var i = 0; i < live.length; i++) {
    liveOk = liveOk && file[bodyAt + iconBufferAt + i] == live[i];
  }
  check(liveOk, 'W8: the live bytes ride at the buffer, at $iconBufferAt');

  // THE PLAN, compiled from this build's own layout against its own rows —
  // the shape a peer's plan is compiled through, with `theirs` the layout on
  // the wire and `mine` this reader's own rows.
  final plan = tables.profileConfigFixedNewPlan();
  check(
    plan.theirs.parse(
      ByteData.sublistView(tables.profileConfigFixedLayout),
      0,
      tables.profileConfigFixedLayout.length,
    ),
    'W8: the layout parses into the plan',
  );
  final made = home.TableFixedCompiler.compile(
    plan,
    tables.profileConfigFixedLayout,
    ByteData.sublistView(tables.profileConfigFixedLayout),
    tables.profileConfigFixedDst,
    tables.profileConfigFixedCover,
    tables.profileConfigFixedCoverCount,
    home.TableFixedReport(),
  );
  check(made > 0, 'W8: the plan compiles ($made entries)');
  home.tableFixedFillRun(
    plan.fill,
    plan.fillCount,
    tables.profileConfigFixedPrefill,
    plan.image,
  );
  final rc = home.TableFixedReport();
  home.tableFixedRun(
    plan.entries,
    made,
    file,
    fileView,
    bodyAt,
    plan.image,
    plan.imageView,
    plan.remap,
    plan.conv,
    rc,
  );
  final back = home.ProfileConfig();
  home.profileConfigFixedDecode(back, plan.image, plan.imageView, 0, rc);
  check(
    back.iconLength == live.length && bytesRide(back.icon),
    'W8 GREEN: the array row lands the buffer in the buffer and the length in '
    'the length — icon reads ${liveText(back.icon, back.iconLength)}',
  );

  // ---- THE NEGATIVE CONTROL: THE TEXT ROW, AS IT WAS ----------------------
  //
  // Swap the two columns back on a COPY of the rows — dst the buffer's offset
  // becomes the length's, and aux the other way — and compile the same plan
  // from the same layout. The count then lands in the buffer's first four
  // bytes and the elements over the length field, and the value must NOT come
  // back: what reports the law is the value that comes back, not a refusal.
  final swapped = Int32List.fromList(tables.profileConfigFixedDst);
  if (bytesRow >= 0) {
    final b = bytesRow * home.TableFixedLane.dstLanes;
    final d = swapped[b + home.TableFixedLane.dstOffset];
    swapped[b + home.TableFixedLane.dstOffset] =
        swapped[b + home.TableFixedLane.dstAux];
    swapped[b + home.TableFixedLane.dstAux] = d;
  }
  final wrongPlan = tables.profileConfigFixedNewPlan();
  check(
    wrongPlan.theirs.parse(
      ByteData.sublistView(tables.profileConfigFixedLayout),
      0,
      tables.profileConfigFixedLayout.length,
    ),
    'W8: the layout parses into the control plan',
  );
  final made2 = home.TableFixedCompiler.compile(
    wrongPlan,
    tables.profileConfigFixedLayout,
    ByteData.sublistView(tables.profileConfigFixedLayout),
    swapped,
    tables.profileConfigFixedCover,
    tables.profileConfigFixedCoverCount,
    home.TableFixedReport(),
  );
  check(
    made2 > 0,
    'W8: the plan compiles either way — the rows are not what refuses '
    '($made2 entries)',
  );
  home.tableFixedFillRun(
    wrongPlan.fill,
    wrongPlan.fillCount,
    tables.profileConfigFixedPrefill,
    wrongPlan.image,
  );
  final rc2 = home.TableFixedReport();
  home.tableFixedRun(
    wrongPlan.entries,
    made2,
    file,
    fileView,
    bodyAt,
    wrongPlan.image,
    wrongPlan.imageView,
    wrongPlan.remap,
    wrongPlan.conv,
    rc2,
  );
  final wrong = home.ProfileConfig();
  home.profileConfigFixedDecode(
    wrong,
    wrongPlan.image,
    wrongPlan.imageView,
    0,
    rc2,
  );
  final right2 = wrong.iconLength == live.length && bytesRide(wrong.icon);
  check(
    !right2,
    'W8 NEGATIVE CONTROL: the text row really does write the count into the '
    'buffer — icon reads ${liveText(wrong.icon, wrong.iconLength)}',
  );

  exit(failures == 0 ? 0 : 1);
}

bool bytesRide(Uint8List icon) {
  for (var i = 0; i < live.length; i++) {
    if (icon[i] != live[i]) {
      return false;
    }
  }
  return true;
}

String liveText(Uint8List icon, int used) {
  final parts = <String>['len $used'];
  for (var i = 0; i < used && i < 4; i++) {
    parts.add('0x${icon[i].toRadixString(16)}');
  }
  return parts.join(' ');
}

File: test/conformance/rust/rows/W8.rs
// W8 — bytes(N) takes the array row (the rust leg, fixed form, form byte 3).
//
// THE LAW (docs/FIXED-FORM-ALGORITHM.md:202, §3.1, fix 10): the writer writes
// `length` units for text and bytes, NEVER all Max — the slack past the live
// extent stays the template's zeros (§3.4). A `bytes(N)` rides the form as
// kind 14, an ARRAY of u8 (§3.4, §4.1 line 696): the count lands at `aux`
// and the elements at `dst`, the same row as a counted array. Text and bytes
// are the two rows the other way round (line 702-703): text is length-at-dst,
// buffer-at-aux; bytes is buffer-at-dst, length-at-aux.
//
// THE PRODUCTION PATH this test drives, end to end, every stage generated:
//   padded_frame_fixed_save     the write of PaddedFrame, which has bytes(12)
//                               blob (tables/block/Padded.schema:42)
//   padded_frame_fixed_load     the read through the IDENTITY plan (this
//                               build's own layout, hash PADDED_FRAME_FIXED_HASH)
//   padded_frame_fixed_write_body / scatter
//                               the per-field stores, named in the generator
//                               (internal/codegen/rusttable/fixedtable.go,
//                               emitWriteBody / emitScatterBody)
//
// THE VECTOR is built HERE from the law: PaddedFrame declares `blob bytes(12)`,
// so the blob takes 4 + 12 = 16 bytes of body. The writer writes the LENGTH
// word at body+3213 and the live bytes at body+3217, exactly `length` units,
// and the slack at body+3221..3229 stays the template's zeros. The file-level
// offsets come from the layout: header 16, layout-len 4, layout 395, record
// hash 8, body 3229 → file offset 423 for the body's first byte.
//
// THE NEGATIVE CONTROL is the wire a wrong writer would produce: the length
// word at the buffer offset and the live bytes at the length offset — the
// text convention, as it was before fix 10. A reader that took the bytes
// row from the layout the SAME way would still land the buffer at its
// declared offset, and the value that comes back is the thing that proves
// the row was an array's, not a text field's. The control keeps every call
// and every include: it is the WRONG WIRE under the SAME reader.
//
//   rustc --edition 2021 --test ... test/conformance/rust/rows/W8.rs \
//       -o build/rows-rust-W8 && ./build/rows-rust-W8
//
// Exit 0 green, 1 red; one printed line per assertion.

use blockdemo::{
    padded_frame_fixed_load, padded_frame_fixed_measure, padded_frame_fixed_save,
    PaddedFrameRow, TableFixedEntry, TableFixedReport, TABLE_FIXED_HEADER_BYTES,
};

static FAILURES: std::sync::atomic::AtomicU32 = std::sync::atomic::AtomicU32::new(0);

fn check(ok: bool, what: &str) {
    if ok {
        println!("ok - {what}");
    } else {
        println!("FAIL - {what}");
        FAILURES.fetch_add(1, std::sync::atomic::Ordering::Relaxed);
    }
}

const LAYOUT_LEN_BYTES: usize = 4;
const RECORD_HASH_BYTES: usize = 8;
// 64 rows × 50 bytes/row + 13 (marker, stamp, count) = 3213
const BODY_BLOB_LENGTH_AT: usize = 3213;
const BODY_BLOB_BUFFER_AT: usize = 3217;
const BODY_BLOB_BUFFER_MAX: usize = 12;

fn file_body_offset() -> usize {
    TABLE_FIXED_HEADER_BYTES + LAYOUT_LEN_BYTES + blockdemo::PADDED_FRAME_FIXED_BLOCK.len()
        + RECORD_HASH_BYTES
}

#[test]
fn test_row_w8() {
    // ---- THE WRITE: one record whose `blob` carries FOUR live bytes ----
    //
    // The writer writes the length word at body+3213, the four live bytes at
    // body+3217..3221, and the eight bytes past them (body+3221..3229) stay
    // the template's zeros — `length` units, never all Max.
    let mut row = PaddedFrameRow::default();
    row.marker = 0x42;
    row.stamp = 0x0102_0304_0506_0708;
    row.rows_count = 0;
    row.blob_length = 4;
    row.blob[0] = 0xDE;
    row.blob[1] = 0xAD;
    row.blob[2] = 0xBE;
    row.blob[3] = 0xEF;

    let measure = padded_frame_fixed_measure(1);
    let mut file = vec![0xABu8; measure];
    let written = padded_frame_fixed_save(&[row], &mut file).expect("the writer saves the record");
    check(written == measure, "W8: the production writer saves the record (returns measure bytes)");

    let body = file_body_offset();

    // THE LENGTH WORD rides at body+3213 (= the body's blob_length offset).
    let len_word = u32::from_le_bytes(
        file[body + BODY_BLOB_LENGTH_AT..body + BODY_BLOB_LENGTH_AT + 4]
            .try_into()
            .expect("four bytes for the length word"),
    );
    check(len_word == 4, "W8: the length word at body+3213 is 4 (the live extent, not Max)");

    // THE LIVE BYTES ride at body+3217 (the body's blob buffer offset).
    let live = &file[body + BODY_BLOB_BUFFER_AT..body + BODY_BLOB_BUFFER_AT + 4];
    check(live == &[0xDE, 0xAD, 0xBE, 0xEF], "W8: the four live bytes ride at body+3217..3221");

    // THE SLACK past the live bytes is the TEMPLATE'S ZEROS — `length` units,
    // never all Max. A writer that took the text row or wrote the full buffer
    // would leave non-zero bytes here, and that is the broken case.
    let mut slack_ok = true;
    for i in BODY_BLOB_BUFFER_AT + 4..BODY_BLOB_BUFFER_AT + BODY_BLOB_BUFFER_MAX {
        if file[body + i] != 0 {
            slack_ok = false;
        }
    }
    check(slack_ok, "W8: the slack past the live bytes is the template's zeros (body+3221..3229)");

    // ---- THE READ: identity plan (this build's own hash) reads it back ---
    let mut values = [PaddedFrameRow::default(); 1];
    let mut plan = [TableFixedEntry::default(); 512];
    let mut remap = [0u16; 512];
    let mut report = TableFixedReport::default();
    let n = padded_frame_fixed_load(&mut values, &file, &mut plan, &mut remap, &mut report)
        .expect("the reader opens its own file");
    check(n == 1, "W8: the reader reads one record");
    let back = &values[0];
    check(back.blob_length == 4, "W8: the reader reads the length word as 4");
    let back_live: [u8; 4] = back.blob[..4].try_into().expect("four bytes");
    check(back_live == [0xDE, 0xAD, 0xBE, 0xEF], "W8: the reader reads the four live bytes back");

    // ---- THE NEGATIVE CONTROL: the wrong wire under the same reader ------
    //
    // A writer that took the TEXT ROW (length at dst, buffer at aux) would
    // produce a wire where the length word lands at the buffer's offset
    // (body+3217) and the live bytes ride at the length's offset (body+3213).
    // Construct that wrong wire by hand — same template, same length, same
    // payload, just the two columns swapped — and the reader takes the bytes
    // ROW from its own layout: it reads the LENGTH word at body+3213
    // (the buffer slot in this wrong wire) and the BUFFER at body+3217
    // (the length slot in this wrong wire). The value that comes back is the
    // thing that proves the row was an array's, not a text field's.
    let mut wrong = file.clone();
    // body+3213..3217 was the length word (= 4); put the live bytes there.
    wrong[body + BODY_BLOB_LENGTH_AT..body + BODY_BLOB_LENGTH_AT + 4]
        .copy_from_slice(&[0xDE, 0xAD, 0xBE, 0xEF]);
    // body+3217..3221 was the live bytes; put the length word there.
    wrong[body + BODY_BLOB_BUFFER_AT..body + BODY_BLOB_BUFFER_AT + 4]
        .copy_from_slice(&4u32.to_le_bytes());

    let mut values = [PaddedFrameRow::default(); 1];
    let mut plan = [TableFixedEntry::default(); 512];
    let mut remap = [0u16; 512];
    let mut report = TableFixedReport::default();
    let n = padded_frame_fixed_load(&mut values, &wrong, &mut plan, &mut remap, &mut report)
        .expect("the reader opens the wrong wire");
    check(n == 1, "W8 NEGATIVE CONTROL: the reader still opens a wire where length and buffer were swapped");
    let back = &values[0];
    let back_live: [u8; 4] = back.blob[..4].try_into().expect("four bytes");
    // Under the wrong wire the reader lands the LENGTH WORD (4 LE: 04 00 00 00)
    // at the buffer offset, so back.blob[0..4] is { 04, 00, 00, 00 } and
    // back.blob_length is the bytes that were at the length slot
    // (0xDEADBEEF clamped to [0, 12] = 12). Neither is the live pattern.
    let live_was_lost = back_live != [0xDE, 0xAD, 0xBE, 0xEF] || back.blob_length != 4;
    check(live_was_lost, "W8 NEGATIVE CONTROL: the text row really did write the count into the buffer — live bytes lost");

    let fails = FAILURES.load(std::sync::atomic::Ordering::Relaxed);
    assert_eq!(fails, 0, "W8: {fails} assertion(s) failed");
}
